macrosynergy/download/dataquery.py

```diff
@@ -285,540 +317,629 @@ def _get_token(self) -> str:
+    :raises <HeartbeatError>: if the heartbeat (check connection) fails.
+    :raises <Exception>: other exceptions may be raised by underlying functions.
+    """
+
+    def __init__(
+        self,
+        oauth: bool = True,
         debug: bool = False,
         concurrent: bool = True,
-        batch_size: int = 20,
-        heartbeat: bool = False,
+        batch_size: int = API_EXPR_LIMIT,
+        heartbeat: bool = True,
+        suppress_warnings: bool = True,
         **kwargs,
     ):
-
         self.proxy = kwargs.pop("proxy", kwargs.pop("proxies", None))
-        self.heartbeat = heartbeat
+        self.heartbeat: bool = heartbeat
         self.msg_errors: List[str] = []
+        self.msg_warnings: List[str] = []
+        self.debug: bool = debug
+        self.concurrent: bool = concurrent
+        self.suppress_warnings: bool = suppress_warnings
+        self.batch_size: int = batch_size
 
         if oauth:
-            self.access: OAuth = OAuth(
-                client_id=kwargs.pop("client_id"),
-                client_secret=kwargs.pop("client_secret"),
-                url=kwargs.pop("base_url", OAUTH_BASE_URL),
-                token_url=kwargs.pop("token_url", OAUTH_TOKEN_URL),
-                dq_resource_id=kwargs.pop("resource_id", OAUTH_DQ_RESOURCE_ID),
+            # ensure that we have a client_id and client_secret
+            try:
+                for k in ["client_id", "client_secret"]:
+                    assert k in kwargs, f"{k} must be provided."
+            except AssertionError as e:
+                raise ValueError(e)
+            except Exception as e:
+                raise e
+
+            self.access_method: OAuth = OAuth(
+                client_id=kwargs["client_id"],
+                client_secret=kwargs["client_secret"],
+                base_url=OAUTH_BASE_URL,
+                token_url=OAUTH_TOKEN_URL,
+                dq_resource_id=OAUTH_DQ_RESOURCE_ID,
                 proxy=self.proxy,
             )
+
         else:
-            self.access: CertAuth = CertAuth(
-                username=kwargs.pop("username"),
-                password=kwargs.pop("password"),
-                crt=kwargs.pop("crt"),
-                key=kwargs.pop("key"),
-                base_url=kwargs.pop("base_url", CERT_BASE_URL),
+            # ensure that we have a username and password, crt and key
+            try:
+                for k in ["username", "password", "crt", "key"]:
+                    assert k in kwargs, f"{k} must be provided."
+            except AssertionError as e:
+                raise ValueError(e)
+            except Exception as e:
+                raise e
+
+            self.access_method: CertAuth = CertAuth(
+                username=kwargs["username"],
+                password=kwargs["password"],
+                crt=kwargs["crt"],
+                key=kwargs["key"],
+                base_url=CERT_BASE_URL,
```

Comment:
Allow for overwrite in the input arguments of default URL path. Internal JPM implementation will require a different URL.